Pluggable congestion control

Ralith opened the pull request:
As discussed in #693. ~Draft until the abstractness of the interface has been validated by the addition of at least one other controller (probably CUBIC/HyStart++).~

djc commented:
I think I'm okay with the config `Arc` now. Can you sketch what your explicit trait alternative for the factory thing would look like? I can't quite wrap my head around what the constraints are.

(Otherwise, things look pretty good to me, though I'll do a final pass once we figure out the factory thing.)

Ralith commented:
The explicit trait approach:

```rust
pub trait CongestionControllerFactory {
    fn new_controller(&self) -> Box<dyn CongestionController>;
}
// ...
impl CongestionControllerFactory for Arc<NewRenoConfig> {
    fn new_controller(&self) -> Box<dyn CongestionController> {
        Box::new(NewReno::new(self.clone(), Instant::now()))
    }
}
```
In use:
```rust
config.congestion_controller_factory(Arc::new(config));
```

djc commented:
How would the `maybe_rebinding` case look for that? I think it'd need a separate impl? How do you feel about current approach vs explicit trait?

Ralith commented:
`(self.config.congestion_controller_factory)()` would be replaced with `self.config.congestion_controller_factory.new_controller()`; no other changes. A trait with one method that takes `&self` is always algebraically equivalent to an analogous `Fn`.

Between the two, I'm leaning towards the explicit trait right now, because `config.congestion_controller_factory(Arc::new(config));` seems like considerably less ceremony than `config.congestion_controller_factory(|| Box::new(NewReno::new(config, Instant::now())))`. It's more boilerplate if you want to invoke complicated custom logic, but that seems like an *extremely* niche case.

djc commented:
Cool, I think I agree.

Ralith commented:
Updated with an explicit factory trait. I also refactored things to make the `congestion` module public, to ease discoverability of different algorithms we add in the future, and limit growth of clutter in the root namespace.

djc commented on quinn-proto/src/congestion.rs:
```diff
@@ -0,0 +1,27 @@
+/// Common interface for different congestion controllers
+pub trait Controller: Send {
+    /// Packet deliveries were confirmed
+    fn on_ack(&mut self, sent: Instant, bytes: u64, congestion_blocked: bool);
```
Hmm, should we pre-emptively add a `now` argument here? That seems like something that might definitely be needed at least conceptually (also nice symmetry with `on_congestion_event`).

Ralith replied:
It's needed by BBR, at least, so I suppose so.

djc commented on quinn-proto/src/congestion.rs:
```diff
@@ -0,0 +1,27 @@
+    fn on_ack(&mut self, sent: Instant, bytes: u64, congestion_blocked: bool);
+
+    /// Packets were deemed lost or marked congested
+    fn on_congestion_event(&mut self, now: Instant, sent: Instant, persistent: bool);
```
The meaning of `persistent` is not obvious to me. Different name? Some comments? Switch from `bool` to `enum`?

Ralith replied:
Elaborated a bit and expanded the name to the full term used in the recovery draft.
